### PR TITLE
[G2M] Dropdown - add dropdownSelect & dropdownSelectSteps Components

### LINKED ISSUE #29: Components Library: New Select Component
# 🎯 Usage 
- Allows user input by selecting from a list of predefined options
- Best used in situations where there are more than 3 options to select from.

# 🎭  Behaviours
- **Display Options**: For any single option, the user can also choose to display an icon and/or a description.
- **Width**: The width of an option element will on default fill the dropdown container. 
- **Sizes**: Two predefined sizes are provided: large and medium.
- **Selection Types**: We offer simple selections (i.e. select process takes only 1 click), and more complex selections such as tree-select, step-select, and linked select (i.e. requires multiple clicks to complete selection and may include multiple dropdown and input fields).
- **Smart Select**: Supports autocomplete and search features.

# 🧩 Style Variations 
- `Simple Select` → A simple select with text options only
- `Grouped Select` → Options in the dropdown are divided by categories/labels
- `Select With Divider` → Options in the dropdown are divided by a divider
- `Select With Icons` → Simple select where each option is displayed with an icon
- `Select With Description` → Simple select where each option is displayed with a description
- `Multi-select` → Select multiple options from the dropdown
- `Tree-select` → A cascade dropdown where options are grouped in a tree-structure
- `Step-select` → Options are grouped and displayed across multiple dropdown steps
- `Linked-select` → Selection that relies on a previous selection to be completed before it becomes available

# 💟 Icons Used
- [check](https://app.streamlinehq.com/icons/streamline-light/interface-essential/form-validation/check-1)
- [arrow-left](https://app.streamlinehq.com/icons/streamline-light/arrows-diagrams/arrows/arrow-left-1)
- [arrow-right](https://app.streamlinehq.com/icons/streamline-light/arrows-diagrams/arrows/arrow-right-1)

# 🔗 Zeplin Link
- [Select](https://zpl.io/ag16Jq0)

# 🔢  Attributes

Property | Type | Default Value | Description
--- | --- | --- | ---
`size` | `large` \| `medium` | `medium` | The size of Select
`defaultValue` | `ReactNode` | `-` | The initial selected option
`placeholder` | `ReactNode` | `-` | The placeholder option shown in the input box
`disabled` | `boolean` | `'false'` | Sets Select to disabled state
`options` | `{ label, value }[]` | `-` | List of selectable options
`dropdownType` | `string` | `-` | Customize the type of dropdown used
`input` | `ReactNode` | `-` | The selected option

# 🔍 Design Preview
Select

### PR DESCRIPTION
**Pull request by kc-leung**
fix #29 
- Added dropdownSelect component with group, divider, icons, description & multiselect variants
- Added dropdownSelectSteps component for steps vartiant
- Added Linked variant in storybook using dropdownSelect component
- Added dropdown-data.js for sample data used in Dropdown storybook
- Added disabled option for dialogBase component
- Added arrow-left & validation-check Icons
- Added additional spacing option `'3px': '0.188rem', '7px': '0.438rem' & '9px': '0.563rem'` to tailwindConfig
- Added fontSize option `'11px': ['0.688rem', { lineHeight: 1.45 }]` to tailwindConfig

TO DO: 
- Update lumen-tree styling for Tree variant

### CODE REVIEW

**Comment by woozyking** [comment]
would recommend arranging the commit message in such fashion:

```
dropdown - add dropdown base component
dropdown - add dropdown select component
dropdown/Base - scrollbar styling & multiselect padding changes
dropdown - add dropdown-select-steps component
dropdown/Select - add linked select variant
dropdown - add each variants props descriptions 
```

so it can be used to auto-generate logically grouped release notes like:

```
dropdown
- add dropdown base component
- add dropdown select component
- add dropdown-select-steps component
- add each variants props descriptions 
- Base
    - scrollbar styling & multiselect padding changes
- Select
    - add linked select variant
```

File: src/base-components/dialog-base.js [hunk 1]
```diff
@@ -26,12 +26,14 @@ const DialogBase = ({ classes, button, children, modal, open, anchor, onClick })
...
+    if (!disabled) {
+      if (!controlledOpen) {
+        setComponentIsActive(!componentIsActive)
+      }
+      onClick()
     }
-    onClick()
   }
-
+
```
Comment (woozyking) [inline src/base-components/dialog-base.js:36]: redundant trailing space

File: src/base-components/dialog-base.js [hunk 2]
```diff
@@ -26,12 +26,14 @@ const DialogBase = ({ classes, button, children, modal, open, anchor, onClick })
   }
 
   const handleClick = () => {
-    if (!controlledOpen) {
-      setComponentIsActive(!componentIsActive)
+    if (!disabled) {
+      if (!controlledOpen) {
+        setComponentIsActive(!componentIsActive)
```
Comment (woozyking) [inline src/base-components/dialog-base.js:31]: theoretically the same, but in practice slightly better/safer:

```js
setComponentIsActive((state) => !state)
```

File: src/components/chip.js [hunk 3]
```diff
@@ -17,7 +17,7 @@ const Chip = forwardRef(({ classes, children, startIcon, endIcon, color, selecta
       focus:outline-none ${borderElementsColor}
       ${selectable ? 'cursor-pointer' : 'pointer-events-none'}
     `, 
-    content: `${classes.content ? classes.content : 'text-xxs font-semibold tracking-lg leading-1.6'}`, 
+    content: `${classes.content ? classes.content : 'text-xxs font-semibold tracking-lg leading-1.6 uppercase'}`,
```
Comment (woozyking) [inline src/components/chip.js:20]: is this supposed to be in uppercase always (by design)?

File: src/components/dropdown-select.js [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+          )
+        })
+      )
+    }  
+
+    return (
+      <>
+        {render}
+      </>
+    )
```
Comment (woozyking) [inline src/components/dropdown-select.js:111]: `return render`

File: src/data/dropdown-data.js [hunk 5]
```diff
@@ -0,0 +1,350 @@
+import React from 'react'
```
Comment (woozyking) [inline src/data/dropdown-data.js:1]: for sample data, better fit them outside of `src/` to not bloat the package size (usually they should reside under `stories/`

File: src/components/chip.js (hunk 3 above)
Comment (kc-leung) [reply src/components/chip.js:20]: yes it is.

**Review by sallkall — CHANGES_REQUESTED** [review]
Looking good!! 🤩

### Step Select
One thing to change for the **step-select** - there should only be one dropdown at any given time. This is my fault I didn't make it clear in the prototype. It made a quick mockup, should behave like this instead:

https://user-images.githubusercontent.com/33563070/128729880-5fdcfbc6-a59e-40c9-8f8b-52b250d076b4.mov


### Dropdown area shadow
There should be a shadow for all dropdown areas like so (see figma/zeplin for details):
image


### Other questions:
1. Category Title - can we enforce **title case** for these? 
2. Mixing options of different styles (i.e. with icon, with description, none etc), not sure if this is already enforced but can we only allow one type of option style at any one given time? So we don't mix option styles? 
3. Weird overflow behaviour - On my 16' mac pro i see this grey square (see photo) that I think is supposed to allow horizontal scroll? But when i click & drag on it nothing happens. It goes away when i zoom in on the page to 110%. Any chance we can fix/get rid of this?
image

**Pull request closed without merging**